addon/mixins/route.js

```diff
@@ -283,17 +278,61 @@ export default Ember.Mixin.create({
 
    @method updateInfinityModel
    @param {Ember.Enumerable} newObjects The new objects to add to the model
-   @return {Ember.Array} returns the updated infinity model
+   @return {DS.RecordArray} returns the updated infinity model
    */
   updateInfinityModel(newObjects) {
-    var infinityModel = this.get(this.get('_modelPath'));
+    const totalPages = newObjects.get(this.get('totalPagesParam'));
+    this.set('_totalPages', totalPages);
 
-    return infinityModel.pushObjects(newObjects.get('content'));
+    let infinityModel = newObjects;
+    if (this.get('_firstPageLoaded')) {
+      infinityModel = this._infinityModel();
+      infinityModel.pushObjects(newObjects.get('content'));
+    }
+
+    this.set('_firstPageLoaded', true);
+    this._notifyInfinityModelUpdated(newObjects);
+
+    const canLoadMore = this.get('_canLoadMore');
+    infinityModel.set('reachedInfinity', !canLoadMore);
+
+    if (!canLoadMore) {
+      this._notifyInfinityModelLoaded();
+    }
+
+    return infinityModel;
   },
 
-  actions: {
-    infinityLoad() {
-      this._infinityLoad();
+  /**
+   notify that the infinity model has been updated
+
+   @private
+   @method _notifyInfinityModelUpdated
+   */
+  _notifyInfinityModelUpdated(newObjects) {
+    if (!this.infinityModelUpdated) {
```

Comment:
`this.infinityModelUpdated` is dead code, this is the only place it's referenced